### PR TITLE
[TS-197] Yaga. Migration library

### PR DESCRIPTION
**Pull request by im-kulikov**
- migrator + tests
- update cli-package
- move testdb to helpers

### CODE REVIEW

File: migrate/helpers.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+// findMigrations in specified folder (path)
+func findMigrations(path string) ([]os.FileInfo, error) {
+	var (
+		err error
+		dir *os.File
+	)
+
+	if dir, err = os.Open(path); os.IsNotExist(err) {
+		return nil, ErrDirNotExists
+	}
```
Comment (jenchik) [inline migrate/helpers.go:60]: no ```dir.Close()```

File: migrate/helpers.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+func CreateMigration(path, name string) error {
+	var (
+		err   error
+		f     *os.File
+		dt    = time.Now().Unix()
+		items = []string{"up", "down"}
+	)
+
+	for _, item := range items {
+		if f, err = os.Create(path + "/" + fmt.Sprintf(fileNameTpl, dt, name, item)); err != nil {
```
Comment (jenchik) [inline migrate/helpers.go]: mb
```
if f, err = os.Create(fmt.Sprintf(filePathTpl, path, dt, name, item)); err != nil {
```

File: migrate/helpers.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	"github.com/go-pg/pg/types"
+)
+
+// getTableName for quote table name
+func getTableName() types.ValueAppender {
+	return pg.Q(tableName)
+}
+
+// extractAttributes, such as  version, name and migration-type
+func extractAttributes(filename string) (version int64, name, mType string, err error) {
```
Comment (jenchik) [inline migrate/helpers.go:24]: mb one regex?

File: migrate/cmd/main.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		PoolSize: 2,
+	}
+}
+
+var (
+	log  = zap.New(zap.Development)
+	opts = new(Options)
+)
+
+func main() {
```
Comment (jenchik) [inline migrate/cmd/main.go:55]: mb added DSN parameter?

**Review by jenchik — DISMISSED** [review]

**Comment by jenchik** [comment]
mb added command ```migrate:list``` ?

File: migrate/cmd/main.go (hunk 4 above)
Comment (jenchik) [reply migrate/cmd/main.go:55]: sorry, but we can not use it

Comment (im-kulikov) [reply migrate/cmd/main.go:55]: yep

File: migrate/helpers.go (hunk 1 above)
Comment (im-kulikov) [reply migrate/helpers.go:60]: fixed

File: migrate/helpers.go (hunk 2 above)
Comment (m1ome) [reply migrate/helpers.go]: `path.Join` guys, it should be cross-os IMO.

Comment (im-kulikov) [reply migrate/helpers.go]: yep..

Comment (im-kulikov) [reply migrate/helpers.go]: fixed

File: migrate/helpers.go (hunk 3 above)
Comment (im-kulikov) [reply migrate/helpers.go:24]: I don't think so

File: helpers/testdb/main.go [hunk 5]
```diff
@@ -24,7 +25,7 @@ type Database struct {
 func GetTestDB() *Database {
 	if database == nil {
 		err := godotenv.Load()
-		if err != nil {
+		if err != nil && !strings.Contains(err.Error(), "no such file or directory") {
```
Comment (m1ome) [inline helpers/testdb/main.go:28]: `const` ?

File: migrate/cmd/main.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	Path     string
+	Type     string
+	Steps    int
+}
+
+func (o Options) Validate() bool {
+	return len(o.Addr) > 0 &&
+		len(o.User) > 0 &&
+		len(o.Database) > 0 &&
+		len(o.Password) > 0 &&
```
Comment (m1ome) [inline migrate/cmd/main.go]: empty password?

File: migrate/cmd/main.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		(o.Type == Up || o.Type == Down || o.Type == List)
+}
+
+func (o Options) PG() *pg.Options {
+	return &pg.Options{
+		Addr:     o.Addr,
+		User:     o.User,
+		Database: o.Database,
+		Password: o.Password,
+		PoolSize: 2,
```
Comment (m1ome) [inline migrate/cmd/main.go]: Maybe use default?

Comment (m1ome) [reply migrate/cmd/main.go]: This looks like a magic number

File: migrate/cmd/main.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+var (
+	log  = zap.New(zap.Development)
+	opts = new(Options)
+)
+
+func main() {
+	flag.StringVar(&opts.Password, "p", "", "password")
+	flag.StringVar(&opts.Addr, "a", "", "address")
```
Comment (m1ome) [inline migrate/cmd/main.go]: `localhost` is default?

File: migrate/cmd/main.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	flag.StringVar(&opts.Database, "d", "", "db name")
+	flag.StringVar(&opts.User, "u", "", "username")
+	flag.StringVar(&opts.Path, "src", "", "path to migrations")
```
Comment (m1ome) [inline migrate/cmd/main.go]: by default `./migrations` ?

File: migrate/helpers.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+	for _, file := range files {
+		// Ignore non sql files:
+		if filepath.Ext(file.Name()) != ".sql" {
+			continue
+		}
+
+		log.Infof("Prepare migration file: %s", file.Name())
+
+		if data, err = ioutil.ReadFile(path + "/" + file.Name()); err != nil {
```
Comment (m1ome) [inline migrate/helpers.go]: path.Join

**Review by m1ome — CHANGES_REQUESTED** [review]

File: migrate/cmd/main.go (hunk 8 above)
Comment (im-kulikov) [reply migrate/cmd/main.go]: hm... it must be `localhost:5432`

File: migrate/cmd/main.go (hunk 7 above)
Comment (im-kulikov) [reply migrate/cmd/main.go]: ok

File: helpers/testdb/main.go (hunk 5 above)
Comment (im-kulikov) [reply helpers/testdb/main.go:28]: there's any sense?

Comment (im-kulikov) [reply helpers/testdb/main.go:28]: it's used just to ignore message `.env no such file or directory`

File: migrate/helpers.go (hunk 3 above)
Comment (jenchik) [reply migrate/helpers.go:24]: did a lot of unnecessary work instead of executing one command

File: migrate/cmd/main.go (hunk 9 above)
Comment (jenchik) [reply migrate/cmd/main.go]: better by default than nothing

File: migrate/helpers.go (hunk 3 above)
Comment (im-kulikov) [reply migrate/helpers.go:24]: I didn't like regexp

File: migrate/vars.go [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+	sqlSelectVersion = `SELECT version, created_at FROM ? ORDER BY id ASC`
+	sqlCreateSchema  = `CREATE SCHEMA IF NOT EXISTS ?`
+	sqlNewVersion    = `INSERT INTO ? (version, created_at) VALUES (?, now())`
+	sqlRemVersion    = `DELETE FROM ? WHERE version = ?`
+	sqlGetVersion    = `SELECT version FROM ? ORDER BY id DESC LIMIT 1`
+	sqlCreateTable   = `
+CREATE TABLE IF NOT EXISTS ? (
+	id serial,
+	version bigint UNIQUE,
```
Comment (m1ome) [inline migrate/vars.go:23]: maybe we should add full migration name?

**Review by m1ome — CHANGES_REQUESTED** [review]

**Review by pcherednichenko — DISMISSED** [review]
`dir.Close()` it is necessary to correct, and looks like everything is ok

______
_Added later:_
aaa, I see, okay

**Review by kpeo — DISMISSED** [review]

**Review by jenchik — APPROVED** [review]

File: migrate/helpers.go (hunk 3 above)
Comment (jenchik) [reply migrate/helpers.go:24]: very bad

**Review by m1ome — APPROVED** [review]

File: migrate/vars.go (hunk 11 above)
Comment (im-kulikov) [reply migrate/vars.go:23]: fixed

**Pull request merged**